clippy_lints/src/undocumented_unsafe_blocks.rs

```diff
@@ -137,10 +137,10 @@ impl UndocumentedUnsafeBlocks {
 
         let between_span = if block_span.from_expansion() {
             self.macro_expansion = true;
-            enclosing_scope_span.with_hi(block_span.hi())
+            enclosing_scope_span.with_hi(block_span.hi()).source_callsite()
         } else {
             self.macro_expansion = false;
-            enclosing_scope_span.to(block_span)
+            enclosing_scope_span.to(block_span).source_callsite()
```

Comment:
This fixes the case in question but I think there are more cases unaccounted for. In general I think two spans should be ensured to have the same `span.ctxt()` before combining them with `to`, `with_hi`, etc, otherwise the operation is unpredictable. There can be any combination of contexts between the two spans.

A general fix might be...

```rust
if enclosing_scope_span.ctxt() != block_span.ctxt() {
    // get the span of the macro call
    // mac! { unsafe {} }
    enclosing_scope_span = hygiene::walk_chain(enclosing_scope_span, block_span.ctxt());
    if enclosing_scope_span.ctxt() != block_span.ctxt() {
        // if the contexts are still not equal after walk_chain,
        // the unsafe block span context is a descendant of the enclosing scope span context
        // { macro_with_unsafe!() }
        // return since I don't know what span to use
        return;
    }
}
```

But then I don't know if this lint is designed to work with the span of a macro call, where there can be an arbitrary amount of code between the start of the macro call and the unsafe block.

Reply:
@flip1995 @giraffate